File: src/ObjectManager/Object.Tes/FilePacks/Records/345-INFO.Dialog Topic Info.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class INFORecord : Record
    {
        // TES3
        public struct DATA3Field
        {
            public int Unknown1;
            public int Disposition;
            public byte Rank; // (0-10)
            public byte Gender; // 0xFF = None, 0x00 = Male, 0x01 = Female
            public byte PCRank; // (0-10)
            public byte Unknown2;

            public DATA3Field(UnityBinaryReader r, int dataSize)
            {
                Unknown1 = r.ReadLEInt32();
                Disposition = r.ReadLEInt32();
                Rank = r.ReadByte();
                Gender = r.ReadByte();
                PCRank = r.ReadByte();
                Unknown2 = r.ReadByte();
            }
        }

        public class TES3Group
        {
            public STRVField NNAM; // Next info ID (form a linked list of INFOs for the DIAL). First INFO has an empty PNAM, last has an empty NNAM.
            public DATA3Field DATA; // Info data
            public STRVField ONAM; // Actor
            public STRVField RNAM; // Race
            public STRVField CNAM; // Class
            public STRVField FNAM; // Faction 
            public STRVField ANAM; // Cell
            public STRVField DNAM; // PC Faction
            public STRVField NAME; // The info response string (512 max)
            public FILEField SNAM; // Sound
            public BYTEField QSTN; // Journal Name
            public BYTEField QSTF; // Journal Finished
            public BYTEField QSTR; // Journal Restart
            public SCPTRecord.CTDAField SCVR; // String for the function/variable choice
            public UNKNField INTV; //
            public UNKNField FLTV; // The function/variable result for the previous SCVR
            public STRVField BNAM; // Result text (not compiled)
        }

        // TES4
        public struct DATA4Field
        {
            public byte Type;
            public byte NextSpeaker;
            public byte Flags;

            public DATA4Field(UnityBinaryReader r, int dataSize)
            {
                Type = r.ReadByte();
                NextSpeaker = r.ReadByte();
                Flags = dataSize == 3 ? r.ReadByte() : (byte)0;
            }
        }

        public class TRDTField
        {
            public uint EmotionType;
            public int EmotionValue;
            public byte ResponseNumber;
            public string ResponseText;
            public string ActorNotes;

            public TRDTField(UnityBinaryReader r, int dataSize)
            {
                EmotionType = r.ReadLEUInt32();
                EmotionValue = r.ReadLEInt32();
                r.SkipBytes(4); // Unused
                ResponseNumber = r.ReadByte();
                r.SkipBytes(3); // Unused
            }

            public void NAM1Field(UnityBinaryReader r, int dataSize)
            {
                ResponseText = r.ReadASCIIString((int)dataSize, ASCIIFormat.PossiblyNullTerminated);
            }

            public void NAM2Field(UnityBinaryReader r, int dataSize)
            {
                ActorNotes = r.ReadASCIIString((int)dataSize, ASCIIFormat.PossiblyNullTerminated);
            }
        }

        public class TES4Group
        {
            public DATA4Field DATA; // Info data
            public FMIDField<QUSTRecord> QSTI; // Quest
            public FMIDField<DIALRecord> TPIC; // Topic
            public List<FMIDField<DIALRecord>> NAMEs = new List<FMIDField<DIALRecord>>(); // Topics
            public List<TRDTField> TRDTs = new List<TRDTField>(); // Responses
            public List<SCPTRecord.CTDAField> CTDAs = new List<SCPTRecord.CTDAField>(); // Conditions
            public List<FMIDField<DIALRecord>> TCLTs = new List<FMIDField<DIALRecord>>(); // Choices
            public List<FMIDField<DIALRecord>> TCLFs = new List<FMIDField<DIALRecord>>(); // Link From Topics
            public SCPTRecord.SCHRField SCHR; // Script Data
            public BYTVField SCDA; // Compiled Script
            public STRVField SCTX; // Script Source
            public List<FMIDField<Record>> SCROs = new List<FMIDField<Record>>(); // Global variable reference
        }

        public override string ToString() => $"INFO: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID - Info name string (unique sequence of #'s), ID
        public FMIDField<INFORecord> PNAM; // Previous info ID
        public TES3Group TES3 = new TES3Group();
        public TES4Group TES4 = new TES4Group();

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            if (formatId == GameFormatId.TES3)
                switch (type)
                {
                    case "INAM": EDID = new STRVField(r, dataSize); DIALRecord.LastRecord?.INFOs.Add(this); return true;
                    case "PNAM": PNAM = new FMIDField<INFORecord>(r, dataSize); return true;
                    case "NNAM": TES3.NNAM = new STRVField(r, dataSize); return true;
                    case "DATA": TES3.DATA = new DATA3Field(r, dataSize); return true;
                    case "ONAM": TES3.ONAM = new STRVField(r, dataSize); return true;
                    case "RNAM": TES3.RNAM = new STRVField(r, dataSize); return true;
                    case "CNAM": TES3.CNAM = new STRVField(r, dataSize); return true;
                    case "FNAM": TES3.FNAM = new STRVField(r, dataSize); return true;
                    case "ANAM": TES3.ANAM = new STRVField(r, dataSize); return true;
                    case "DNAM": TES3.DNAM = new STRVField(r, dataSize); return true;
                    case "NAME": TES3.NAME = new STRVField(r, dataSize); return true;
                    case "SNAM": TES3.SNAM = new FILEField(r, dataSize); return true;
                    case "QSTN": TES3.QSTN = new BYTEField(r, dataSize); return true;
                    case "QSTF": TES3.QSTF = new BYTEField(r, dataSize); return true;
                    case "QSTR": TES3.QSTR = new BYTEField(r, dataSize); return true;
                    case "SCVR": TES3.SCVR = new SCPTRecord.CTDAField(r, dataSize, formatId); return true;
                    case "INTV": TES3.INTV = new UNKNField(r, dataSize); return true;
                    case "FLTV": TES3.FLTV = new UNKNField(r, dataSize); return true;
                    case "BNAM": TES3.BNAM = new STRVField(r, dataSize); return true;
                    default: return false;
                }
            switch (type)
            {
                case "DATA": TES4.DATA = new DATA4Field(r, dataSize); return true;
                case "QSTI": TES4.QSTI = new FMIDField<QUSTRecord>(r, dataSize); return true;
                case "TPIC": TES4.TPIC = new FMIDField<DIALRecord>(r, dataSize); return true;
                case "NAME": TES4.NAMEs.Add(new FMIDField<DIALRecord>(r, dataSize)); return true;
                case "TRDT": TES4.TRDTs.Add(new TRDTField(r, dataSize)); return true;
                case "NAM1": ArrayUtils.Last(TES4.TRDTs).NAM1Field(r, dataSize); return true;
                case "NAM2": ArrayUtils.Last(TES4.TRDTs).NAM2Field(r, dataSize); return true;
                case "CTDA":
                case "CTDT": TES4.CTDAs.Add(new SCPTRecord.CTDAField(r, dataSize, formatId)); return true;
                case "TCLT": TES4.TCLTs.Add(new FMIDField<DIALRecord>(r, dataSize)); return true;
                case "TCLF": TES4.TCLFs.Add(new FMIDField<DIALRecord>(r, dataSize)); return true;
                case "SCHR":
                case "SCHD": TES4.SCHR = new SCPTRecord.SCHRField(r, dataSize); return true;
                case "SCDA": TES4.SCDA = new BYTVField(r, dataSize); return true;
                case "SCTX": TES4.SCTX = new STRVField(r, dataSize); return true;
                case "SCRO": TES4.SCROs.Add(new FMIDField<Record>(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-INFO.Dialog Topic Info.swift
//
//  INFORecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class INFORecord: Record {
    // TES3
    public struct DATA3Field {
        public let unknown1: Int32
        public let disposition: Int32
        public let rank: UInt8 // (0-10)
        public let gender: UInt8 // 0xFF = None, 0x00 = Male, 0x01 = Female
        public let pcRank: UInt8 // (0-10)
        public let unknown2: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            unknown1 = r.readLEInt32()
            disposition = r.readLEInt32()
            rank = r.readByte()
            gender = r.readByte()
            pcRank = r.readByte()
            unknown2 = r.readByte()
        }
    }

    public class TES3Group {
        public var NNAM: STRVField!  // Next info ID (form a linked list of INFOs for the DIAL). First INFO has an empty PNAM, last has an empty NNAM.
        public var DATA: DATA3Field! // Info data
        public var ONAM: STRVField!  // Actor
        public var RNAM: STRVField!  // Race
        public var CNAM: STRVField!  // Class
        public var FNAM: STRVField!  // Faction
        public var ANAM: STRVField!  // Cell
        public var DNAM: STRVField!  // PC Faction
        public var NAME: STRVField!  // The info response string (512 max)
        public var SNAM: FILEField!  // Sound
        public var QSTN: BYTEField!  // Journal Name
        public var QSTF: BYTEField!  // Journal Finished
        public var QSTR: BYTEField!  // Journal Restart
        public var SCVR: SCPTRecord.CTDAField! // String for the function/variable choice
        public var INTV: UNKNField! //
        public var FLTV: UNKNField! // The function/variable result for the previous SCVR
        public var BNAM: STRVField! // Result text (not compiled)
    }

    // TES4
    public struct DATA4Field {
        public let type: UInt8
        public let nextSpeaker: UInt8
        public let flags: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            type = r.readByte()
            nextSpeaker = r.readByte()
            flags = dataSize == 3 ? r.readByte() : 0
        }
    }

    public class TRDTField {
        public let emotionType: UInt32
        public let emotionValue: Int32
        public let responseNumber: UInt8
        public var responseText: String = ""
        public var actorNotes: String = ""

        init(_ r: BinaryReader, _ dataSize: Int) {
            emotionType = r.readLEUInt32()
            emotionValue = r.readLEInt32()
            r.skipBytes(4) // Unused
            responseNumber = r.readByte()
            r.skipBytes(3) // Unused
        }

        func NAM1Field(_ r: BinaryReader, _ dataSize: Int) {
            responseText = r.readASCIIString(dataSize, format: .possibleNullTerminated)
        }

        func NAM2Field(_ r: BinaryReader, _ dataSize: Int) {
            actorNotes = r.readASCIIString(dataSize, format: .possibleNullTerminated)
        }
    }

    public class TES4Group {
        public var DATA: DATA4Field! // Info data
        public var QSTI: FMIDField<QUSTRecord>! // Quest
        public var TPIC: FMIDField<DIALRecord>! // Topic
        public var NAMEs = [FMIDField<DIALRecord>]() // Topics
        public var TRDTs = [TRDTField]() // Responses
        public var CTDAs = [SCPTRecord.CTDAField]() // Conditions
        public var TCLTs = [FMIDField<DIALRecord>]() // Choices
        public var TCLFs = [FMIDField<DIALRecord>]() // Link From Topics
        public var SCHR: SCPTRecord.SCHRField! // Script Data
        public var SCDA: BYTVField! // Compiled Script
        public var SCTX: STRVField! // Script Source
        public var SCROs = [FMIDField<Record>]() // Global variable reference
    }

    public override var description: String { return "INFO: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID - Info name string (unique sequence of #'s), ID
    public var PNAM: FMIDField<INFORecord>! // Previous info ID
    public var TES3 = TES3Group()
    public var TES4 = TES4Group()

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        if format == .TES3 {
            switch type {
            case "INAM": EDID = r.readSTRV(dataSize); DIALRecord.lastRecord?.INFOs.append(self)
            case "PNAM": PNAM = FMIDField<INFORecord>(r, dataSize)
            case "NNAM": TES3.NNAM = r.readSTRV(dataSize)
            case "DATA": TES3.DATA = DATA3Field(r, dataSize)
            case "ONAM": TES3.ONAM = r.readSTRV(dataSize)
            case "RNAM": TES3.RNAM = r.readSTRV(dataSize)
            case "CNAM": TES3.CNAM = r.readSTRV(dataSize)
            case "FNAM": TES3.FNAM = r.readSTRV(dataSize)
            case "ANAM": TES3.ANAM = r.readSTRV(dataSize)
            case "DNAM": TES3.DNAM = r.readSTRV(dataSize)
            case "NAME": TES3.NAME = r.readSTRV(dataSize)
            case "SNAM": TES3.SNAM = r.readSTRV(dataSize)
            case "QSTN": TES3.QSTN = r.readT(dataSize)
            case "QSTF": TES3.QSTF = r.readT(dataSize)
            case "QSTR": TES3.QSTR = r.readT(dataSize)
            case "SCVR": TES3.SCVR = SCPTRecord.CTDAField(r, dataSize, format)
            case "INTV": TES3.INTV = r.readBYTV(dataSize)
            case "FLTV": TES3.FLTV = r.readBYTV(dataSize)
            case "BNAM": TES3.BNAM = r.readSTRV(dataSize)
            default: return false
            }
            return true
        }
        switch type {
        case "DATA": TES4.DATA = DATA4Field(r, dataSize)
        case "QSTI": TES4.QSTI = FMIDField<QUSTRecord>(r, dataSize)
        case "TPIC": TES4.TPIC = FMIDField<DIALRecord>(r, dataSize)
        case "NAME": TES4.NAMEs.append(FMIDField<DIALRecord>(r, dataSize))
        case "TRDT": TES4.TRDTs.append(TRDTField(r, dataSize))
        case "NAM1": TES4.TRDTs.last!.NAM1Field(r, dataSize)
        case "NAM2": TES4.TRDTs.last!.NAM2Field(r, dataSize)
        case "CTDA",
             "CTDT": TES4.CTDAs.append(SCPTRecord.CTDAField(r, dataSize, format))
        case "TCLT": TES4.TCLTs.append(FMIDField<DIALRecord>(r, dataSize))
        case "TCLF": TES4.TCLFs.append(FMIDField<DIALRecord>(r, dataSize))
        case "SCHR",
             "SCHD": TES4.SCHR = SCPTRecord.SCHRField(r, dataSize)
        case "SCDA": TES4.SCDA = r.readBYTV(dataSize)
        case "SCTX": TES4.SCTX = r.readSTRV(dataSize)
        case "SCRO": TES4.SCROs.append(FMIDField<Record>(r, dataSize))
        default: return false
        }
        return true
    }
}
